src/odm_sharing/private/trees.py

```diff
@@ -0,0 +1,457 @@
+'''
+This module contains the parsing of a set of rules into an abstract syntax tree.
+
+- the list of input rules can only reference other preceeding rules
+- a parsing context (ctx) stores all previously created nodes
+- a node has a type/kind, a value, and a list of children
+- the children of a node is called 'sons' since it's shorter
+- nodes are first added to ctx and then later added to parent nodes with O(1)
+  lookup, this way the tree is constructed incrementally while parsing each
+  rule
+- share nodes are made to be children of a single root-node, since each org
+  gets its own node, there may be multiple share-rules, and the tree
+  can only have a single root node
+- the root-node is updated every time a new share-node is added
```

Comment:
I can't update nodes since they're immutable, I can only replace them. I like that they're immutable since it can prevent bugs and forces the code to be more functional and streamlined.